calc/lib/calc/parser.ex

```diff
@@ -0,0 +1,162 @@
+defmodule Calc.Parser do
+  @moduledoc """
+  Module with functions to parse given expression into evaluation tree
+  """
+
+  defguardp is_op(op) when is_atom(op) and op in [:add, :mul, :div, :sub, :lparen, :rparen]
+
+  @doc """
+  Parses given string with expression and returns operation tree
+
+  ### Examples
+
+      iex>  Calc.Parser.parse("2 * 3 + 4")
+      {:ok, {:add, {:mul, {:num, 2.0}, {:num, 3.0}}, {:num, 4.0}}}
+
+      iex(3)> Calc.Parser.parse("-2 + 3 / 4")
+      {:ok, {:add, {:neg, {:num, 2.0}}, {:div, {:num, 3.0}, {:num, 4.0}}}}
+
+  """
+  def parse(exp) do
+    with {:ok, tokens} <- tokenize(exp),
+         {:ok, tree} <- rd_parser(tokens) do
+      {:ok, tree}
+    end
+  end
+
+  @doc """
+  Tokenize given string, convert all numbers to floats
+
+  ### Examples
+
+      iex> Calc.Parser.tokenize("14 + 21 - 13 + 43 * 55")
+      {:ok, [14.0, :add, 21.0, :sub, 13.0, :add, 43.0, :mul, 55.0]}
+
+      iex> Calc.Parser.tokenize("asdf + 43")
+      {:error, "failed to parse token asdf: not a number"}
+
+      iex> Calc.Parser.tokenize("12a + 43")
+      {:error, "failed to parse token 12a: extra chars"}
+  """
+  def tokenize(exp) do
+    exp
+    |> to_charlist
+    |> split([])
+    |> reverse_and_convert([])
+  end
+
+  # Splits given iolist to tokens
+  # Returns reversed list of tokens, and each "string" in that list
+  # is also reversed. Need to pass thru reverse_and_convert fo final result
+  defp split([], acc), do: acc
+  defp split([?\s | tail], acc), do: split(tail, acc)
+  defp split([?+ | tail], acc), do: split(tail, [:add | acc])
+  defp split([?- | tail], acc), do: split(tail, [:sub | acc])
+  defp split([?/ | tail], acc), do: split(tail, [:div | acc])
+  defp split([?* | tail], acc), do: split(tail, [:mul | acc])
+  defp split([?\( | tail], acc), do: split(tail, [:lparen | acc])
+  defp split([?\) | tail], acc), do: split(tail, [:rparen | acc])
```

Comment:
Мне лень проверять, но, вроде бы, эскейпить ничего не нужно.
```suggestion
  defp split([?( | tail], acc), do: split(tail, [:lparen | acc])
  defp split([?) | tail], acc), do: split(tail, [:rparen | acc])
```

Reply:
Да, не нужно. меня форматтер в vscode смутил, он не очень справляется с этими скобками.